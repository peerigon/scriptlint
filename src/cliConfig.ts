import commander from "commander";
// eslint-disable-next-line @typescript-eslint/no-var-requires
const { version } = require("../package.json");

type CliConfig = {
	packageFile?: string;
	fix?: boolean;
	strict?: boolean;
	json?: boolean;
	config?: boolean;
};

export default (argv: Array<string>): CliConfig => {
	const program = new commander.Command();
	let packageFile;

	program
		.version(`${version}`)
		.arguments("[packageFile]")
		.action((arg: string) => {
			packageFile = arg;
		})
		.option("-s, --strict", "strict mode")
		.option("-j, --json", "JSON output")
		.option("-c, --config", "inspect the config")
		.option("-f, --fix", "autofixing");

	program.parse(argv);

	// eslint-disable-next-line @typescript-eslint/no-unnecessary-condition
	const cliConfig: CliConfig = { packageFile: packageFile ?? process.cwd() };

	if (program.fix !== undefined) {
		cliConfig.fix = program.fix;
	}

	if (program.strict !== undefined) {
		cliConfig.strict = program.strict;
	}

	if (program.json !== undefined) {
		cliConfig.json = program.json;
	}

	if (program.config !== undefined) {
		cliConfig.config = program.config;
	}

	return cliConfig;
};
