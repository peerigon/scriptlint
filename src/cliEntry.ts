import cliRun from "./cli";

cliRun(process.argv);
