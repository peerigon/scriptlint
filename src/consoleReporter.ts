import chalk from "chalk";
import { Values, MessageBuffer, MessageType } from "./types";
import { makeMessage } from "./utils";

const PREFIX = "⧙։⧘";
let stashed: MessageBuffer = [];

const stash = (message: string, type: MessageType): void => {
	stashed.push({
		message,
		type
	});
};

export const warning = (template: string, values?: Values): void => {
	const message = makeMessage(template, values);

	stash(message, "warning");
};

export const success = (template: string, values?: Values): void => {
	const message = makeMessage(template, values);

	stash(message, "success");
};

export const dump = (jsonOutput: boolean): number => {
	const problemCount = stashed.length;

	if (jsonOutput) {
		// eslint-disable-next-line no-console
		console.log(JSON.stringify(stashed, null, "\t"));
	} else {
		stashed.forEach(({ message, type }) => {
			print(type, message);
		});
	}

	stashed = [];

	return problemCount;
};

export const error = (message: string): void => {
	print("error", message);
};

const print = (type: MessageType, message: string) => {
	switch (type) {
	case "error": {
		// eslint-disable-next-line no-console
		console.log(chalk.bold.red(`${PREFIX} [error] ${message}`));
		break;
	}

	case "warning": {
		const notUnderlined = chalk.yellow.bold(`${PREFIX} [${type}]`);
		const underlined = chalk.yellow.underline(message);

		// eslint-disable-next-line no-console
		console.log(`${notUnderlined} ${underlined}`);
		break;
	}
	case "success": {
		// eslint-disable-next-line no-console
		console.log(chalk.bold.gray(`${PREFIX} [✔️] ${message}`));
		break;
	}
	}
};
