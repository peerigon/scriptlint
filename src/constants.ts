import { Config } from "./types";

export const PROCESS_EXIT_ERROR = 1;
export const PROCESS_EXIT_OK = 0;

export const DEFAULT_CONFIG: Config = {
	strict: false,
	fix: false,
	json: false,
	config: false,
	packageFile: undefined,
	rules: {},
	customRules: [],
	ignoreScripts: []
};

export const PROJECT_NAME = "scriptlint";

export const NAME_REGEX = /^[\d:A-Za-z]+$/;

export const NAMESPACES = [
	"build",
	"dev",
	"format",
	"other",
	"report",
	"setup",
	"start",
	"test"
];

export const DEFAULT_NPM_HOOKS = [
	"install",
	"postinstall",
	"postpack",
	"postpublish",
	"postrestart",
	"postshrinkwrap",
	"poststart",
	"poststop",
	"posttest",
	"postuninstall",
	"postversion",
	"preinstall",
	"precommit",
	"prepack",
	"prepare",
	"prepublish",
	"prepublishOnly",
	"prerestart",
	"preshrinkwrap",
	"prestart",
	"prestop",
	"pretest",
	"preuninstall",
	"preversion",
	"publish",
	"release",
	"restart",
	"shrinkwrap",
	"stop",
	"uninstall",
	"version"
];
