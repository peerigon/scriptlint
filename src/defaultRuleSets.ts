import defaultRules from "./rules";
import { Rule } from "./types";

export const optionalRules = ["natural-order"];

const strict = defaultRules
	.map((r: Rule) => r.name)
	.filter((r) => !optionalRules.includes(r));

const ruleSets = {
	default: [
		"mandatory-test",
		"mandatory-start",
		"mandatory-dev",
		"no-default-test",
	],
	strict,
	optionalRules,
};

export default ruleSets;
