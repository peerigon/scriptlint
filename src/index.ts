/*
路---------路
| {scr:pt |
|   l:nt} |
|         |
|         |
路---------路
*/

import module from "./module";

export = module;
