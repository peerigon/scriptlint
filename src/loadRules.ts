import defaultRuleSets, {optionalRules} from "./defaultRuleSets";
import defaultRules from "./rules";
// Types
import { Rule } from "./types";

type RulesConfig = {
	[key: string]: boolean;
};

export const getRuleByName = (
	rules: Array<Rule>,
	name: string
): Rule | null => {
	const filtered = rules.filter((r: Rule) => r.name === name);

	if (filtered.length < 1) {
		return null;
	}

	return filtered[0];
};

const loadDefaultRulesFromSet = (strict: boolean): Array<Rule> => {
	if (strict) {
		return defaultRuleSets.strict
			.map((name: string) => getRuleByName(defaultRules, name))
			.filter((r): r is Rule => r !== null);
	}

	return defaultRuleSets.default
		.map((name: string) => getRuleByName(defaultRules, name))
		.filter((r): r is Rule => r !== null);
};

const loadOptionalRules = (): Array<Rule> => {
	return optionalRules
		.map((name: string) => getRuleByName(defaultRules, name))
		.filter((r): r is Rule => r !== null);
};

export const loadRulesFromRuleConfig = (
	strict: boolean,
	rulesConfig?: RulesConfig,
	customRules?: Array<Rule>
): Array<Rule> => {
	const optionalRules = loadOptionalRules();
	const rules = loadDefaultRulesFromSet(strict);

	// standard rulesets apply
	if (!rulesConfig) {
		return rules;
	}

	// there's custom rules loaded
	const loadedCustomRules = customRules
		? customRules.filter((cr: Rule) => rulesConfig[cr.name])
		: [];

	// there's enabled optional rules
	const enabledOptionalRules = optionalRules.filter(
		({ name }) => name in rulesConfig && Boolean(rulesConfig[name])
	);

	return [...loadedCustomRules, ...enabledOptionalRules, ...rules].filter(
		({ name }) => !(name in rulesConfig) || rulesConfig[name] !== false
	);
};
