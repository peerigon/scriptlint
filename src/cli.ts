#!/usr/bin/env node
import loadUserConfig from "./userConfig";
import loadCliConfig from "./cliConfig";
import { makePackageFilePath } from "./utils";
import userPackageScriptContext from "./userPackageScripts";
import { loadRulesFromRuleConfig } from "./loadRules";
import execute from "./execute";
import { success, warning, dump, error } from "./consoleReporter";
import {
	DEFAULT_CONFIG,
	PROCESS_EXIT_ERROR,
	PROCESS_EXIT_OK
} from "./constants";

/* istanbul ignore next */
const processExit = (code: number) => {
	if (typeof process.env.JEST_WORKER_ID === "undefined") {
		// eslint-disable-next-line no-process-exit
		process.exit(code);
	}
};

export default (argv: Array<string>) => {
	try {
		/**
		 * config assembly
		 */

		const userConfig = loadUserConfig();
		const cliConfig = loadCliConfig(argv);
		const config = {
			...DEFAULT_CONFIG,
			...{ json: false },
			...userConfig,
			...cliConfig
		};

		// output the config (--config) but only if we don't want JSON output on the CLI
		if (!config.json && config.config) {
			const json = JSON.stringify(config, null, 2);

			// eslint-disable-next-line no-console
			console.log(`\n\n${json}\n\n`);
		}

		/**
		 * package.json
		 */

		const {
			writePackageScripts,
			readPackageScripts
		} = userPackageScriptContext(
			makePackageFilePath(config.packageFile ?? process.cwd())
		);

		const scripts = readPackageScripts(config.ignoreScripts);

		/**
		 * rule loading
		 */

		const rules = loadRulesFromRuleConfig(
			config.strict,
			config.rules,
			config.customRules
		);

		/**
		 * and go!
		 */
		const [issues, fixedScripts, issuesFixed] = execute(
			rules,
			scripts,
			warning,
			config.fix
		);

		if (config.fix && issuesFixed > 0) {
			writePackageScripts(fixedScripts);
			success(`Fixed ${issuesFixed} issue${issuesFixed > 1 ? "s" : ""}!`);
		}

		if (issues.length - issuesFixed === 0) {
			success("âœ¨  All good");

			dump(config.json);

			processExit(PROCESS_EXIT_OK);
		}

		dump(config.json);
		processExit(PROCESS_EXIT_ERROR);
	} catch (err) {
		error(err);
	}
};
