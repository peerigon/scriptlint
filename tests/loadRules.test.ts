import { loadRulesFromRuleConfig, getRuleByName } from "../src/loadRules";
import defaultRules from "../src/rules";
import {optionalRules} from "../src/defaultRuleSets";

describe("loadRules.ts", () => {
	const defaultRulesLoaded = loadRulesFromRuleConfig(false);
	const strictRulesLoaded = loadRulesFromRuleConfig(true);

	it("loads custom rules", () => {
		const customRule = {
			name: "test-custom-rule-foobar",
			isObjectRule: false,
			message: "foobar",
			validate: () => true,
		};

		const rulesWithCustomRule = loadRulesFromRuleConfig(
			false,
			{
				[customRule.name]: true,
			},
			[customRule]
		);

		expect(
			rulesWithCustomRule
				.map((r) => r.name)
				.filter((r) => r === customRule.name)[0]
		).toBe(customRule.name);
	});

	it("loads optional rules", () => {
		const rulesWithCustomRule = loadRulesFromRuleConfig(
			false,
			{
				"natural-order": true,
			}
		);

		expect(rulesWithCustomRule.map(r => r.name)[0]).toBe("natural-order");
	});

	it("loads correct amount of rules", () => {
		expect(strictRulesLoaded.length).toBe(defaultRules.length - optionalRules.length);
	});

	test("getRuleByName() nulls on unknown name", () => {
		expect(getRuleByName(defaultRulesLoaded, "foo")).toBe(null);
		expect(
			getRuleByName(defaultRulesLoaded, "mandatory-dev")
		).toMatchSnapshot();
	});

	test("loadRulesFromSet() excludes rules by config", () => {
		const rules = loadRulesFromRuleConfig(true, {
			"mandatory-dev": false,
			"mandatory-start": false,
			"mandatory-test": false,
		});

		expect(rules[0].name).toEqual("no-default-test");
	});

	test("loadRulesFromSet() adds custom rules", () => {
		const rules = loadRulesFromRuleConfig(
			true,
			{
				foobarbaz: true,
			},
			[
				{
					name: "foobarbaz",
					isObjectRule: true,
					message: "barbazfoo",
					validate: () => true,
				},
			]
		);

		expect(rules[0].name).toEqual("foobarbaz");
	});

	test("loadRulesFromSet() ignores custom rules that are not loaded", () => {
		const rules = loadRulesFromRuleConfig(false, {}, [
			{
				name: "foobarbaz",
				isObjectRule: true,
				message: "barbazfoo",
				validate: () => true,
			},
		]);

		expect(rules.map((r) => r.name).includes("foobarbaz")).toBe(false);
	});
});
