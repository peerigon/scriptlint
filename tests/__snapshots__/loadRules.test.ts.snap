// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`loadRules.ts getRuleByName() nulls on unknown name 1`] = `
Object {
  "isObjectRule": true,
  "message": "must contain a \\"dev\\" script",
  "name": "mandatory-dev",
  "validate": [Function],
}
`;
